src/nts_ke/server/connection.rs

```diff
@@ -200,13 +234,64 @@ impl KeServerConn {
 
             let keys = gen_key(&self.tls_session).unwrap();
 
+            while !self.ntske_state.finished {
+                // need to read 4 bytes to get the header.
+                if reader.len() < HEADER_SIZE {
+                    info!(self.logger, "readable nts-ke stream is not enough to read header");
+                    self.ntske_buffer = Vec::from(reader);
+                    return;
+                }
+
+                // need to read the body_length to get the body.
+                let body_length = u16::from_be_bytes([reader[2], reader[3]]) as usize;
+                if reader.len() < HEADER_SIZE + body_length {
+                    info!(self.logger, "readable nts-ke stream is not enough to read body");
+                    self.ntske_buffer = Vec::from(reader);
+                    return;
+                }
+
+                // Reconstruct the whole record byte array to let the `records` module deserialize it.
+                let mut record_bytes = vec![0; HEADER_SIZE + body_length];
+                reader.read_exact(&mut record_bytes).unwrap();
```

Comment:
This seems a bit dangerous. I know we've checked we've got enough length, but panics aren't great.